BaristaSwift/Common/Common/Base/Store.swift

```diff
@@ -89,95 +88,87 @@ public class Store<objectType: StoreProtocol> {
+            .then { _ -> Promise<objectType> in
+                if let record = self.record(forExternalId: entry.externalId) {
+                    return Promise.value(record)
+                } else {
+                    return Promise(error:StoreErrors.recordNotFound)
+                }
+        }
     }
     
-    public func locallyUpdateEntry<T:StoreProtocol>(entry: T) {
-        var record: T = entry
-        record.local = true
-        record.locallyUpdated = true
-        self.upsertEntries(record: record)
+    public func updateEntry(entry: objectType) -> Promise<objectType> {
+        _ = locallyUpdateEntry(entry: entry)
+        return syncUpDown()
+            .then { _ -> Promise<objectType> in
+                if let record = self.record(forExternalId: entry.externalId) {
+                    return Promise.value(record)
+                } else {
+                    return Promise(error:StoreErrors.recordNotFound)
+                }
+        }
     }
     
-    public func updateEntry<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
-        record.local = true
-        record.locallyUpdated = true
-        syncEntry(entry: record, completion: completion)
+    public func deleteEntry(entry: objectType) -> Promise<Void> {
+        locallyDeleteEntry(entry: entry)
+        return syncUpDown()
     }
 
-    public func syncEntry<T:StoreProtocol>(entry: T, completion: SyncCompletion = nil) {
-        var record: T = entry
-        record.objectType = T.objectName
-        store.upsertEntries([record.data], toSoup: T.objectName)
-        syncUp() { syncState in
-            if let _ = syncState?.isDone() {
-                self.syncDown(completion: completion)
-            }
-        }
+    public func syncEntry(entry: objectType) -> Promise<Void> {
+        var record: objectType = entry
+        record.objectType = objectType.objectName
+        store.upsertEntries([record.data], toSoup: objectType.objectName)
+        return syncUpDown()
     }
 
-    public func syncDown(completion: SyncCompletion = nil) {
-        smartSync.reSync(byName: syncDownName, update: completion ?? { _ in return })
-    }
-    
-    public func syncUp(completion: SyncCompletion = nil) {
-        let updateBlock: SFSyncSyncManagerUpdateBlock = { [unowned self] (syncState: SFSyncState?) in
-            if let syncState = syncState {
-                if syncState.isDone() || syncState.hasFailed() {
-                    DispatchQueue.main.async {
-                        if syncState.hasFailed() {
-                            SalesforceSwiftLogger.log(type(of:self), level:.error, message:"syncUp \(objectType.objectName) failed")
-                        }
-                    }
-                    completion?(syncState)
+    private func reSync(syncName: String) -> Promise<Void> {
+        return smartSync.Promises.reSync(syncName: syncName)
+            .then { syncState -> Promise<Void> in
```

Comment:
@trooper2013 I think we should change our SmartSync extensions to have the sync promises resolve when the sync is done and reject when the sync has failed. That way, I wouldn't to do what I did here. Getting a failed sync state in the next then complicates the code of the callers.